firedrake/ufl_expr.py

```diff
@@ -138,15 +138,24 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
 
     See also :func:`ufl.derivative`.
     """
-    if set(extract_coefficients(form)) & set(u.split()):
+    # TODO: What about Constant?
+    if len(u.split()) > 1 and set(extract_coefficients(form)) & set(u.split()):
         raise ValueError("Taking derivative of form wrt u, but form contains coefficients from u.split()."
                          "\nYou probably meant to write split(u) when defining your form.")
     if du is None:
-        if isinstance(u, function.Function):
+        if isinstance(u, firedrake.Function):
             V = u.function_space()
             args = form.arguments()
             number = max(a.number() for a in args) if args else -1
             du = Argument(V, number + 1)
+        elif isinstance(u, firedrake.Constant):
+            if u.ufl_shape != ():
+                raise ValueError("Real function space of vector elements not supported")
+            mesh = form.ufl_domain()
+            V = firedrake.FunctionSpace(mesh, "Real", 0)
+            args = form.arguments()
+            number = max(a.number() for a in args) if args else -1
+            du = Argument(V, number + 1)
```

Comment:
Should be able to differentiate wrt `Constant`.  Again, suggests `Constant` should just be `Function(FunctionSpace(mesh, "Real", 0))`.